replace all colon characters for label generation

boehm107 opened the pull request:
Previously, it was possible to create services with invalid label values that contained one or more colons, e.g.

sslcertificates/issuer: this:is:my:issuer

This adjustment removes all colons in the $cert_issuer variable. When searching for such service labels, a crash report is no longer generated.

boehm107 committed:
use sed to replace all colon ':' characters in cert_issuer

Previously, it was possible to create services with invalid label values that contained one or more colons, e.g.

sslcertificates/issuer: this:is:my:issuer

This adjustment removes all colons in the $cert_issuer variable. When searching for such service labels, a crash report is no longer generated.

gurubert commented:
This should be addressed at the label generation, not within the agent.

boehm107 commented:
> This should be addressed at the label generation, not within the agent.

Hello, I completely agree with you. But that was the quickest solution to avoid invalid labels.  😅

I have thought about this again and have two solutions in the `agent_based` plug-in. It could look like this:

1. Solution for type `issuer` only:
```python
if data.get('issuer'):
 sl.append(ServiceLabel(u'sslcertificates/issuer', data['issuer'].replace(":", ""))))
```

2. Solution for all labels:
```python
def discover_sslcertificates(params, section: SSLCertificatesSection) -> DiscoveryResult:
    def cleanup_label(value):
        if isinstance(value, str):
            return value.replace(":", "")
        return value

    label_map = {
        'issuer_hash': 'sslcertificates/issuer_hash',
        'issuer': 'sslcertificates/issuer',
        'algosign': 'sslcertificates/algorithm',
        'template': 'sslcertificates/template',
    }

    for name, data in section.items():
        if 'min_lifetime' in params and 'starts' in data:
            if data['expires'] - data['starts'] < params['min_lifetime']:
                continue

        sl = []
        for key, label in label_map.items():
            val = data.get(key)
            if val:
                sl.append(ServiceLabel(label, cleanup_label(val)))
        yield Service(item=name, labels=sl)
```

I have already successfully tested both solutions with version 2.3.0p34 and 2.4.0p4.

With the approach for all labels I am not sure if there are fields/types in SSL certificates that contain a “colon”. But for Checkmk it would be the better way, as otherwise a crash report would be generated when searching for such corresponding labels.

![image](https://github.com/user-attachments/assets/1b98a564-7c57-4947-9869-94d915a1c850)

gurubert commented:
I would prefer your second solution.

boehm107 commented:
> I would prefer your second solution.

This is also my favorite solution. 😊

gurubert commented:
> > I would prefer your second solution.
> 
> This is also my favorite solution. 😊

Then please change the pull request.

boehm107 committed:
Merge branch 'HeinleinSupport:cmk2.3' into cmk2.3

boehm107 committed:
Revert Changes

This should be addressed at the label generation.

boehm107 committed:
Replace all colon characters for label generation

This solution removes all colon characters for label generation in the discover_sslcertificates function in the agent_based plug-in. The function has also been slightly modified for this purpose.

boehm107 commented:
> > > I would prefer your second solution.
> > 
> > 
> > This is also my favorite solution. 😊
> 
> Then please change the pull request.

I have modified and changed the pull request. I hope it all fits.
